app/Livewire/TaStatsGeneral.php

```diff
@@ -0,0 +1,127 @@
+<?php
+
+namespace App\Livewire;
+
+use App\TrainingTicket;
+use Carbon\Carbon;
+use Livewire\Component;
+
+class TaStatsGeneral extends Component {
+    public $no_shows;
+    public $tickets_by_certs;
+    public $session_ids;
+    public $completion_ratios;
+    public \StdClass $lookback_monthly;
+    public \StdClass $lookback_annual;
+    private $cert_types = ['S1', 'S2', 'S3', 'C1'];
+    public $position_select;
+
+    public function render() {
+        $this->generate_stats();
+        return view('livewire.ta-stats-general');
+    }
+
+    private function generate_stats(): void {
+        $this->generate_no_shows();
+        $this->generate_tickets_by_certification();
+        $this->generate_completion_ratios();
+        $this->generate_lookback();
+    }
+
+    private function generate_no_shows(): void {
+        $by_controller = [];
+        $no_show_sessions = TrainingTicket::where('type', 10)->where('date', '<=', Carbon::now()->addDays(45)->toDateTimeString())->get();
+        foreach ($no_show_sessions as $no_show_session) {
+            $by_controller[] = $no_show_session->controller_id;
+        }
+        $more_than_one_no_show = array_count_values($by_controller);
+        if (count($more_than_one_no_show) == 0) {
+            $this->no_shows = [];
+            return;
+        }
+        if (max($more_than_one_no_show) < 2) {
+            $this->no_shows = [];
+            return;
+        }
```

Comment:
We could set `$this->no_shows = []` earlier on as a default (line 33). Then combine these into one if

Reply:
I combined the `$this->no_shows = [];` on line 33 per your suggestion to eliminate the repetitive line. Unfortunately, the two if statements are required. We first have to check if the array has more than zero elements, then we can check max(). If you try max() on an array with zero elements, you'll throw an error. Running count() first ensures that this block of code won't attempt to run max() before we've validated that it is permissible to do so.

Reply:
Changing this to `if (count($more_than_one_no_show) == 0 || max($more_than_one_no_show) < 2) {...}` (order does matter here) will cause the if statement to short circuit if the first condition is `true`. So it'll never evaluate the second statement if the first statement is `true`. This speeds up long if statements and allows an easy way to combine if statements.

https://stackoverflow.com/questions/5694733/does-php-have-short-circuit-evaluation

Reply:
How did I not know this????

Reply:
Code updated to reflect this magic